Added data availability checks (#1566)
ethereum/eth2.0-specs: specs/core/2_data-availability-proofs.md

        domain=get_domain(state, DOMAIN_DATA_AVAILABILITY_SLASH, get_current_epoch(state))
    )
    # Get generalized indices (for proof verification)
    generalized_indices = [get_generalized_index(DataAvailabilityProof, 'row_cutoff')]
    if proof.is_column:
        # Case 1: proof is getting indices along a column
        assert proof.checking_extension is False
        generalized_indices.append(get_generalized_index(DataAvailabilityProof, 'columns', proof.axis_index))
        coordinates = [(index, proof.axis_index) for index in proof.indices]
    elif proof.checking_extension:
        # Case 2: proof is getting indices along a row, we are checking against the extension root
        generalized_indices.append(get_generalized_index(DataAvailabilityProof, 'extension', proof.axis_index - CHUNKS_PER_ROW))
        coordinates = [(proof.axis_index, index) for index in proof.indices]
    else:
        # Case 3: proof is getting indices along a row, we are checking against the row root
        generalized_indices.append(get_generalized_index(DataAvailabilityProof, 'rows', proof.axis_index))
        coordinates = [(proof.axis_index, index) for index in proof.indices]

>     # Each individual element can be come from the row/extension trees, or from the column tree

Author: ```suggestion
    # Each individual element can be come from the row/extension trees, or from the column tree
    assert proof.indices == sorted(set(proof.indices))
```

> # Ethereum 2.0 Phase 2 -- Data Availability Proofs

Author: ```suggestion
# Ethereum 2.0 Phase 2 -- Data Availability Checks
```

**Notice**: This document is a work-in-progress for researchers and implementers.

## Table of contents

<!-- TOC -->

> - [Ethereum 2.0 Phase 2 -- Data Availability Proofs](#ethereum-20-phase-2----data-availability-proofs)

Author: ```suggestion
- [Ethereum 2.0 Phase 2 -- Data Availability Checks](#ethereum-20-phase-2----data-availability-checks)
```

    - [Table of contents](#table-of-contents)
    - [Introduction](#introduction)
    - [Constants](#constants)
        - [Misc](#misc)
    - [Data structures](#data-structures)
        - [`DataExtensionSlashing`](#dataextensionslashing)
    - [Helper functions](#helper-functions)
        - [`badd`](#badd)
        - [`bmul`](#bmul)
        - [`eval_polynomial_at`](#eval-polynomial-at)
        - [`interpolate`](#interpolate)
        - [`fill`](#fill)
        - [`fill_axis`](#fill-axis)
        - [`get_data_sqiare`](#get-data-square)
        - [`extend_data_square`](#extend-data-square)
        - [`mk_data_root`](#mk-data-root)
        - [`process_data_extension_slashing`](#process-data-extension-slashing)

<!-- /TOC -->

## Introduction

> This document describes the expected formula for calculating data availability proofs and the beacon chain changes (namely, slashing conditions) needed to enforce them.

Author: ```suggestion
This document describes the expected formula for calculating the Merkle root used for data availability checks and the beacon chain changes (namely, slashing conditions) needed to enforce them.
```
